sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,621 @@
+    delay(30);
+  if (cnt <= 0) {
+    klog("ERROR: failed to get stable clock");
+    return ETIMEDOUT;
+  }
+
+  return 0;
+}
+
+#define CMD_TYPE_SUSPEND 0x00400000
+#define CMD_TYPE_RESUME 0x00800000
+#define CMD_TYPE_ABORT 0x00c00000
+#define CMD_DATA_TRANSFER 0x00200000
+#define CMD_DATA_READ 0x00000010
+#define CMD_DATA_MULTI 0x00000020
+#define CMD_CHECKCRC 0x00080000
+#define CMD_CHECKIDX 0x00100000
+#define CMD_RESP136 0x00010000
+#define CMD_RESP48 0x00020000
+#define CMD_RESP48B 0x00030000
+
+/* This function might be (and probably is!) incomplete, but it does enough
+ * to handle the current block device scenario */
+static uint32_t encode_cmd(emmc_cmd_t cmd) {
+  uint32_t code = (uint32_t)cmd.cmd_idx << 24;
+
+  switch (cmd.exp_resp) {
+    case EMMCRESP_NONE:
+      break;
+    case EMMCRESP_R2:
+      code |= CMD_RESP136;
+      break;
+    case EMMCRESP_R1B:
+      code |= CMD_RESP48B;
+      break;
+    default:
+      code |= CMD_RESP48;
+  }
+  switch (emmc_cmdtype(cmd.flags)) {
+    case EMMC_F_SUSPEND:
+      code |= CMD_TYPE_SUSPEND;
+      break;
+    case EMMC_F_RESUME:
+      code |= CMD_TYPE_RESUME;
+      break;
+    case EMMC_F_ABORT:
+      code |= CMD_TYPE_ABORT;
+      break;
+    default:
+      break;
+  }
+  if (cmd.flags & EMMC_F_DATA_READ) {
+    code |= CMD_DATA_TRANSFER | CMD_DATA_READ;
+  }
+  if (cmd.flags & EMMC_F_DATA_WRITE) {
+    code |= CMD_DATA_TRANSFER;
+  }
+  if (cmd.flags & EMMC_F_DATA_MULTI) {
+    code |= CMD_DATA_MULTI;
+  }
+  if (cmd.flags & EMMC_F_CHKIDX)
+    code |= CMD_CHECKIDX;
+  if (cmd.flags & EMMC_F_CHKCRC)
+    code |= CMD_CHECKCRC;
+
+  return code;
+}
+
+static int bcmemmc_set_bus_width(bcmemmc_state_t *state, uint32_t bw) {
+  switch (bw) {
+    case EMMC_BUSWIDTH_1:
+      b_clr(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_8BIT);
+      b_clr(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_DWITDH);
+      return 0;
+    case EMMC_BUSWIDTH_4:
+      b_clr(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_8BIT);
+      b_set(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_DWITDH);
+      return 0;
+    case EMMC_BUSWIDTH_8:
+      b_set(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_8BIT);
+      b_clr(state->emmc, BCMEMMC_CONTROL0, C0_HCTL_DWITDH);
+      return 0;
+    default:
+      return EINVAL;
+  }
+}
+
+static int bcmemmc_get_bus_width(bcmemmc_state_t *state) {
+  uint32_t ctl = b_in(state->emmc, BCMEMMC_CONTROL0);
+  if (ctl & C0_HCTL_8BIT)
+    return EMMC_BUSWIDTH_8;
+  if (ctl & C0_HCTL_DWITDH)
+    return EMMC_BUSWIDTH_4;
+  return EMMC_BUSWIDTH_1;
+}
+
+static int bcmemmc_get_prop(device_t *cdev, uint32_t id, uint64_t *var) {
+  assert(cdev->parent && cdev->parent->driver == &bcmemmc_driver);
+  bcmemmc_state_t *state = (bcmemmc_state_t *)cdev->parent->state;
+  resource_t *emmc = state->emmc;
+
+  uint32_t reg = 0;
+  switch (id) {
+    case EMMC_PROP_RW_BLKCNT:
+      reg = b_in(emmc, BCMEMMC_BLKSIZECNT);
+      *var = (reg & 0xffff0000) >> 16;
+      return 0;
```

Comment:
Sure, I can do that, but how would that decrease the amount and density of the code?

Reply:
I didn't mean to change the default case (it's ok). The change wouldn't decrease the code density, but it seems more natural to see a break after each case instead of a return (I was clearly referring to the density of "returns").

Reply:
It doesn't seem more natural to me, but whatever. I replaced the returns with breaks.